benchmark/preconditioner/preconditioner.cpp

```diff
@@ -157,41 +158,32 @@ void run_preconditioner(const char *precond_name,
             for (auto i = 0u; i < FLAGS_warmup; ++i) {
                 precond->generate(system_matrix)->apply(lend(b), lend(x_clone));
             }
+            auto generate_timer = get_timer(exec, FLAGS_gpu_timer);
+            auto apply_timer = get_timer(exec, FLAGS_gpu_timer);
 
             exec->synchronize();
-            auto g_tic = std::chrono::steady_clock::now();
-
+            generate_timer->tic();
             std::unique_ptr<gko::LinOp> precond_op;
             for (auto i = 0u; i < FLAGS_repetitions; ++i) {
                 precond_op = precond->generate(system_matrix);
             }
-
-            exec->synchronize();
-            auto g_tac = std::chrono::steady_clock::now();
+            generate_timer->toc();
 
             auto generate_time =
-                std::chrono::duration_cast<std::chrono::nanoseconds>(g_tac -
-                                                                     g_tic) /
-                FLAGS_repetitions;
+                generate_timer->get_total_time() / FLAGS_repetitions;
             add_or_set_member(this_precond_data["generate"], "time",
-                              generate_time.count(), allocator);
+                              generate_time, allocator);
 
             exec->synchronize();
-            auto a_tic = std::chrono::steady_clock::now();
-
+            apply_timer->tic();
             for (auto i = 0u; i < FLAGS_repetitions; ++i) {
                 precond_op->apply(lend(b), lend(x_clone));
             }
+            apply_timer->toc();
 
-            exec->synchronize();
-            auto a_tac = std::chrono::steady_clock::now();
-
-            auto apply_time =
-                std::chrono::duration_cast<std::chrono::nanoseconds>(a_tac -
-                                                                     a_tic) /
-                FLAGS_repetitions;
-            add_or_set_member(this_precond_data["apply"], "time",
-                              apply_time.count(), allocator);
+            auto apply_time = apply_timer->get_total_time() / FLAGS_repetitions;
```

Comment:
Since this is outside the loop and the timer doesn't know about `FLAGS_repetitions`, wouldn't this give a wrong result? (equivalent to `get_total_time()`).

Reply:
The `apply_timer` does know how often it was started and stopped, therefore, it is independent of the `FLAGS_repetitions` (and in my opinion safer) when using `get_average_time()`.

Reply:
https://github.com/ginkgo-project/ginkgo/pull/669/files#diff-d573d9686e53b5ce4b41de8061f46d1a30693d2dfa59795bec0680f6bdd4e4dcR178-R183

Reply:
It's independent only if you put it inside the for loop, AFAIK it is not here. See the context of the line highlighted:
https://github.com/ginkgo-project/ginkgo/pull/669/files#diff-d573d9686e53b5ce4b41de8061f46d1a30693d2dfa59795bec0680f6bdd4e4dcR178-R183

Reply:
We put the timer out of for-loop, so the apply_timer can not know the number of for-loop. As terry said, I keep the current timer for less overhead. If we need to refill/reset the x,b like spmv or solver, we can put the timer inside the for-loop